subprojects/docs/src/docs/userguide/lazyConfiguration.adoc

```diff
@@ -15,40 +15,206 @@
+    <sourcefile file="build.gradle"/>
+    <output args="consumer"/>
+</sample>
+++++
+
+In the example above, the task output and input are connected before any location is defined. This is possible because the input and output properties use the `Provider` API. The output property is created with api:org.gradle.api.DefaultTask#newOutputFile()[] and the input property is created with api:org.gradle.api.DefaultTask#newInputFile()[]. Values are only resolved when they are needed during execution. The setters can be called at any time before the task is executed and the change will automatically affect all related input and output properties.
+
+Another thing to note is the absence of any explicit task dependency. Properties created via `newOutputFile()` and `newOutputDirectory()` bring knowledge about which task is generating them, so using them as task input will implicitly link tasks together.
 
 [[sec:lazy_configuration_faqs]]
 === FAQs
+* Should I expose the api:org.gradle.api.provider.Property[] directly?
+** Yes. For configurable properties (subtypes of api:org.gradle.api.provider.Property[]), there should be a single getter that returns the api:org.gradle.api.provider.Property[]. For non-configurable properties, there should be a single getter that returns a api:org.gradle.api.provider.Provider[].
+* Should I encapsulate the api:org.gradle.api.provider.Property[] behind setter/getter?
+** No. The api:org.gradle.api.provider.Property[] and api:org.gradle.api.provider.Provider[] types have all of the overloads you need to query or configure a value.
+* How can I simplify calls like `obj.getProperty().get()` and `obj.getProperty().set(T)`?
+** You don’t. This makes it very explicit when a property's value is resolved. Generally, you only call api:org.gradle.api.provider.Provider#get()[] in a task action. In most cases, you pass around lazy representations instead.
+* How will the Provider API affect new and existing Gradle tasks?
+** The Provider API will be used for new properties going forward. The Groovy Gradle DSL adds convenience methods to make the use of Providers mostly transparent in build scripts.
+** Existing tasks will have their existing "raw" properties replaced by Providers as needed and in a backwards compatible way.
+** New tasks will be designed with the Provider API.
+* How should I migrate my plugin to use providers?
+** If it's a new property, expose it as a api:org.gradle.api.provider.Property[] or api:org.gradle.api.provider.Provider[] using a single getter.
+** If it's incubating, change it to use a api:org.gradle.api.provider.Property[] or api:org.gradle.api.provider.Provider[] using a single getter.
+** If it's a stable property, add a new api:org.gradle.api.provider.Property[] or api:org.gradle.api.provider.Provider[] and deprecate the old one. You should wire the old getter/setters into the new property as appropriate.
+
```

Comment:
One last thing, I think we should add a short section + sample for using `ListProperty` since it's different enough from the others.

Reply:
I added another section to explain `ListProperty`.